contents/docs/hog/index.mdx

```diff
@@ -284,6 +286,47 @@ Hog's standard library includes the following functions and will expand. To see
 - `sha256Hex(arg: string): string`
 - `sha256HmacChainHex(arg: string[]): string`
 
+## Why Hog?
+
+Truth be told, we didn't set out to build our own programming language. We actually wanted to build the **HogVM**: a way to run HogQL expressions synchronously within our NodeJS ingestion pipeline. However, things escalated quickly.
+
+### HogVM
+
+The HogVM is a stack-based bytecode virtual machine (VM). It was built to power event filtering in our real-time ingestion flow.
+
+It's paired with a compiler that converts HogQL expressions like `properties.$screen_width < properties.$screen_height` into bytecode like `["_H", 1, 32, "$screen_height", 32, "properties", 1, 2, 32, "$screen_width", 32, "properties", 1, 2, 15, 35]`. The HogVM then executes that bytecode.
+
+The HogVM had a few design constraints:
+
+- Virtually no startup time (thousands of different expressions must run against millions of events every minute).
+- Fast enough to run hundreds (or thousands?) of these events per second on a single core.
+- Secure enough to run user provided code.
+- Syntax compatible with HogQL.
+
+Effectively, we wanted to run HogQL synchronously within a NodeJS process, at scale. So that's what we built, twice.
+
+We now have two identical HogVMs: one in [Python](https://github.com/PostHog/posthog/blob/master/hogvm/python/execute.py) (to run alongside queries, unreleased), and one in [TypeScript](https://github.com/PostHog/posthog/blob/master/hogvm/typescript/src/execute.ts) (to run alongside ingestion). We might even build a third version in Rust soon. The surface area for the VM is really that small.
```

Comment:
You touch on it here but I'd put it explicitly in the list above - this VM is small and simple, by design, to make putting it in lots of things easy (my eventual goal would be a single implementation, in rust, that ships node and python bindings, rather than 3 impls, but that's a discussion for next quarter).